python311Packages.horizon-eda: init at 2.5.0

jue89 opened the pull request:
## Description of changes

This PR introduces the Python interface for [Horizon EDA](https://github.com/horizon-eda/horizon).

Usage is described briefly in the [Horizon Docs](https://horizon-eda.readthedocs.io/en/latest/python.html).

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pbsds commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -45,14 +48,36 @@ stdenv.mkDerivation rec {
+    cp build/horizon.so $out/${python3Packages.python.sitePackages}
+  '';
+
+  pythonImportsCheck = lib.optionals (pythonSupport) [
+    "horizon"
   ];
```
For this to run you likely have to include `python3Packages.pythonImportsCheckHook` as a native build input. Verify it works by changing it to something erroneous. ;)

jue89 replied:
Thank you for the hint! My patch was very try-and-error-driven, tbh :D

jue89 replied:
I tried that, but import wasn't possible. Maybe my `installPhase` is too hacky?

pbsds reviewed (changes requested):
Cool approach, but is it possible to lessen the build burden by reusing `horizon-eda` in `python3Packages.horizon-eda`?

---

Result of `nixpkgs-review pr 250633` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>horizon-eda</li>
    <li>python310Packages.horizon-eda</li>
    <li>python311Packages.horizon-eda</li>
  </ul>
</details>

jue89 commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -45,14 +48,36 @@ stdenv.mkDerivation rec {
+    python3
+  ];
+
+  postBuild = lib.optionalString (pythonSupport) ''
+    make -j$NIX_BUILD_CORES pymodule
+  '';
```
Thank you for your feedback :)

Good point: Compiling just the `pymodule` target consumes way less time. I wasn't aware of that. I try to modify the derivation accordingly.

pbsds commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -1,75 +1,33 @@
 stdenv.mkDerivation rec {
-  pname = "horizon-eda";
-  version = "2.5.0";
+  inherit (base) pname version src meta CASROOT;
```
With this formulation you're not adding yourself as a maintainer. Unless @guserav want to take upon himself the maintainership I'd assume you intended to add yourself?

jue89 replied:
I've added myself to the maintainers list

pbsds commented on pkgs/applications/science/electronics/horizon-eda/base.nix:
```diff
@@ -0,0 +1,57 @@
+    sha256 = "sha256-UcjbDJR6shyETpanNkRoH8LF8r6gFjsyNHVSCMHKqS8=";
+  };
+
+  meta = with lib; {
+    description = "A free EDA software to develop printed circuit boards";
+    homepage = "https://horizon-eda.org";
+    maintainers = with maintainers; [ guserav ];
+    license = licenses.gpl3;
+    platforms = platforms.linux;
+  };
```
Please move the `meta` block to the bottom of the file

jue89 replied:
Done.

pbsds commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -1,75 +1,33 @@
     boost.dev
-    pkg-config
     wrapGAppsHook
+    python3
```
Is this needed by the non-python package?

jue89 replied:
Yes, it's using Python during build-time.

pbsds reviewed (changes requested):
Promising! Some more nits from my end. Also, before this can be merged we have to squash the fixup commits. In nixpkgs it is common to force-push to the PR branch

jue89 commented on pkgs/applications/science/electronics/horizon-eda/module.nix:
```diff
@@ -0,0 +1,40 @@
+    enableOSMesa = true;
+  };
+in
+stdenv.mkDerivation {
```
Hmm, I had problems as buildPythonPackage tries to use setuptools. It seems that horizon-eda is pretty uncommon, as it just uses plain Makefiles to build the module. Or maybe I'm holding it wrong?

jue89 replied:
I've read some code and manged to use `buildPythonPackage` :)

jue89 commented:
> Promising! Some more nits from my end. Also, before this can be merged we have to squash the fixup commits. In nixpkgs it is common to force-push to the PR branch

Thank you very much for your review! Currently I'm a few days off. Once I got back home, I'll apply your suggestions.

jue89 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4845,6 +4845,8 @@ self: super: with self; {
 
   hopcroftkarp = callPackage ../development/python-modules/hopcroftkarp { };
 
+  horizon-eda = toPythonModule (callPackage ../applications/science/electronics/horizon-eda/module.nix { });
```
Moved the python module. That's your intention of this suggestion, right?

jue89 commented on pkgs/applications/science/electronics/horizon-eda/base.nix:
```diff
@@ -0,0 +1,57 @@
+, sqlite
+}:
+
+rec {
```
Done

jue89 commented:
I think I applied all your suggestions. `nix run nixpkgs#nixpkgs-review pr 250633` also looks like to work without complaints.

(And sorry @marsam for pinging ... I accidentally added an unrelated change. Already fixed that.)

pbsds reviewed (changes requested):
Some more nits, then it's looking good.

---

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

---


Result of `nixpkgs-review pr 250633` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>horizon-eda</li>
    <li>python310Packages.horizon-eda</li>
    <li>python311Packages.horizon-eda</li>
  </ul>
</details>

jue89 commented on pkgs/applications/science/electronics/horizon-eda/base.nix:
```diff
@@ -0,0 +1,47 @@
+  CASROOT = pkgs.opencascade-occt;
+
+  buildInputs = [
+    pkgs.cppzmq
+    pkgs.curl
+    pkgs.glm
+    pkgs.gtkmm3
+    pkgs.libarchive
+    pkgs.libepoxy
+    pkgs.libgit2
+    pkgs.librsvg
+    pkgs.libuuid
+    pkgs.opencascade-occt
+    pkgs.podofo
+    pkgs.sqlite
```
I realized that `pkgs.mesa` and `mesa` aren't the same when the python module is built. The `callPackage` provided by *python-packages.nix* seems to include all python packages in the top-scope and the python module of `mesa` overrides the package `mesa`. Or am I misunderstanding the concept how `callPackage` works?

jue89 replied:
Thanks for the hint! Looks like it works like a charm :)

pbsds commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4993,6 +4993,8 @@ self: super: with self; {
 
   hopcroftkarp = callPackage ../development/python-modules/hopcroftkarp { };
 
+  horizon-eda = callPackage ../development/python-modules/horizon-eda { };
```
```suggestion
  horizon-eda = callPackage ../development/python-modules/horizon-eda { 
    inherit (pkgs) mesa;
  };
```

jue89 committed:
maintainers: add jue89

jue89 committed:
python311Packages.horizon-eda: init at 2.5.0

jue89 commented on pkgs/development/python-modules/horizon-eda/default.nix:
```diff
@@ -0,0 +1,49 @@
+}:
+
+let
+  base = callPackage ../../../applications/science/electronics/horizon-eda/base.nix { };
```
I exposed the `base` object itself instead of the path. Should work as well, right?

jue89 commented:
Thank you @pbsds :) I really appreciate all the work you put into my PR.

pbsds commented:
@ofborg build

jue89 commented:
Is there anything I could do to drive further reviewing forward?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1165

wegank commented on pkgs/applications/science/electronics/horizon-eda/base.nix:
```diff
@@ -0,0 +1,58 @@
+    owner = "horizon-eda";
+    repo = "horizon";
+    rev = "v${version}";
+    sha256 = "sha256-UcjbDJR6shyETpanNkRoH8LF8r6gFjsyNHVSCMHKqS8=";
```
```suggestion
    hash = "sha256-UcjbDJR6shyETpanNkRoH8LF8r6gFjsyNHVSCMHKqS8=";
```

wegank commented on pkgs/applications/science/electronics/horizon-eda/base.nix:
```diff
@@ -0,0 +1,58 @@
+    sha256 = "sha256-UcjbDJR6shyETpanNkRoH8LF8r6gFjsyNHVSCMHKqS8=";
+  };
+
+  CASROOT = opencascade-occt;
```
```suggestion
  env.CASROOT = opencascade-occt;
```

wegank commented on pkgs/applications/science/electronics/horizon-eda/base.nix:
```diff
@@ -0,0 +1,58 @@
+    description = "A free EDA software to develop printed circuit boards";
+    homepage = "https://horizon-eda.org";
+    maintainers = with maintainers; [ guserav jue89 ];
+    license = licenses.gpl3;
```
`gpl3Only` or `gpl3Plus`?

wegank commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -1,75 +1,38 @@
 
+let
+  base = callPackage ./base.nix { };
+in
 stdenv.mkDerivation rec {
-  pname = "horizon-eda";
-  version = "2.5.0";
+  inherit (base) pname version src meta CASROOT;
```
Could you just add a flag `enablePython ? false` and then unify the two packages under that flag in the same file? Then you can just

```nix
  horizon-eda = toPythonModule (pkgs.horizon-eda.override {
    enablePython = true;
    python3 = python;
  });
```

in `pkgs/top-level/python-packages.nix`.

jue89 replied:
I did that in the first place but @pbsds and I discovered that building the binary together with the python package is a huge overhead in terms of build time. That's the reason for the split.

Aside from that a single file becomes hard to read because of the required dependencies very significantly based on whether the binary or the python module is built.

Or do you think we should go the route with the crowded file?

pbsds replied:
Crowded files are quite common.A bunch of `++ lib.optionals enablePython [...]` could be the way to go.
 
The important difference is the make flags, which can be conditioned on `enablePython`.
Considering the base horizon-eda is now _not_ built when building the python module, the flag `enablePython` is IMO better named `buildPythonModule`, as it also disables the normal build.

jue89 replied:
Okay, I'll change that.

(For me as a newbie those crowded files are super frustrating. It's hard to get started with nixpkgs. This PR already consumed days of work just because it's hard to figure out where things are coming from and how they work. And I'm guesstimating stuff till this day :D)

jue89 replied:
I hope I understood your suggestion correctly? Let me know if this needs further adjustments.

jue89 committed:
horizon-eda: clarify package license

Source: https://github.com/horizon-eda/horizon/blob/v2.5.0/org.horizon_eda.HorizonEDA.metainfo.xml#L6

wegank commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -29,47 +32,70 @@ stdenv.mkDerivation rec {
     "PREFIX="
   ];
 
+  installPhase = lib.optionalString (buildPythonPackage) ''
+    runHook preInstall
+
+    mkdir -p $out/${python3Packages.python.sitePackages}
+    cp build/horizon.so $out/${python3Packages.python.sitePackages}
+
+    runHook postInstall
+  '';
```
I'm not sure about this expression, it looks that `installPhase` is empty for the non-Python package.

jue89 replied:
> I'm not sure about this expression, it looks that `installPhase` is empty for the non-Python package.

If the `installPhase` is empty, it should pick the default one provided by the builder, or am I wrong? At least I'm able to build the binary and the produced `result` has everything I expect to be present.

Currently, horizon-eda has no `install` target for the python module. We can't get around ugly solutions, I guess?

> I'd suggest putting most of the logic in `// lib.optionalAttrs buildPythonPackage { ... }` after meta, or we just go back to [2a97676](https://github.com/NixOS/nixpkgs/commit/2a97676bc1f456b0409344383b463d003c425eb9). What do you think?

Idk. I'm a newbie and it's hard for me to tell which one is the better solution. I would do every change you ask for if this helps to bring this one upstream.

wegank replied:
`lib.optionalString false "..."` would evaluate to the empty string instead of `null`, so I don't expect this would work.

wegank commented on pkgs/applications/science/electronics/horizon-eda/default.nix:
```diff
@@ -1,22 +1,25 @@
 { stdenv
+, lib
+, buildPythonPackage ? false
```
Not sure if it's a good name, since `python3.pkgs.buildPythonPackage` exists.

jue89 replied:
`buildPythonModule`?

wegank commented:
@ofborg build horizon-eda python311Packages.horizon-eda

jue89 commented:
Thank you a lot for your review and support, @wegank @pbsds! I really appreciate it.
